internal/provider/providers/scaleway/provider.go

```diff
@@ -0,0 +1,191 @@
+		field_name: providerSpecificSettings.FieldName,
+		ttl:		providerSpecificSettings.TTL,
+	}, nil
+}
+
+func validateSettings(domain, secretKey, fieldType string) (err error) {
+	err = utils.CheckDomain(domain)
+	if err != nil {
+		return fmt.Errorf("%w: %w", errors.ErrDomainNotValid, err)
+	}
+
+	switch {
+    case secretKey == "":
+        return fmt.Errorf("%w", errors.ErrSecretKeyNotSet)
+    }
+	
+	if fieldType != "A" && fieldType != "AAAA" {
+		return fmt.Errorf("field_type can only be A or AAAA. %w: %s", errors.ErrBadRequest, fieldType)
+	}
+
+	return nil
+}
+
+func (p *Provider) String() string {
+	return utils.ToString(p.domain, p.owner, constants.Dyn, p.ipVersion)
+}
+
+func (p *Provider) Domain() string {
+	return p.domain
+}
+
+func (p *Provider) Owner() string {
+	return p.owner
+}
+
+func (p *Provider) IPVersion() ipversion.IPVersion {
+	return p.ipVersion
+}
+
+func (p *Provider) IPv6Suffix() netip.Prefix {
+	return p.ipv6Suffix
+}
+
+func (p *Provider) Proxied() bool {
+	return false
+}
+
+func (p *Provider) BuildDomainName() string {
+	return utils.BuildDomainName(p.owner, p.domain)
+}
+
+func (p *Provider) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain: fmt.Sprintf("<a href=\"http://%s\">%s</a>", p.BuildDomainName(), p.BuildDomainName()),
+		Owner:  p.Owner(),
+		Provider:  "<a href=\"https://www.scaleway.com/\">Scaleway</a>",
+		IPVersion: p.ipVersion.String(),
+	}
+}
+
+// Update updates the DNS record for the domain using Scaleway's API.
+// API documentation: https://www.scaleway.com/en/developers/api/domains-and-dns/#path-records-update-records-within-a-dns-zone
+func (p *Provider) Update(ctx context.Context, client *http.Client, ip netip.Addr) (newIP netip.Addr, err error) {
+    // Construct the URL for the API request
+    u := url.URL{
+        Scheme: "https",
+        Host:   "api.scaleway.com",
+        Path:   fmt.Sprintf("/domain/v2beta1/dns-zones/%s/records", p.domain),
+        RawQuery: fmt.Sprintf("A=%s", ip.String()),
+    }
+
+    // Prepare the request body
+    requestBody := map[string]interface{}{
+        "changes": []map[string]interface{}{
+            {
+                "set": map[string]interface{}{
+                    "id_fields": map[string]interface{}{
+                        "name": p.field_name,
+                        "type": p.field_type,
+                    },
+                    "records": []map[string]interface{}{
+                        {
+                            "data": ip.String(),
+                            "ttl":  p.ttl,
+                        },
+                    },
+                },
+            },
+        },
+    }
+    requestBodyBytes, err := json.Marshal(requestBody)
+    if err != nil {
+        return netip.Addr{}, fmt.Errorf("json marshal: %w", err)
+    }
+
+    // Create the HTTP request
+    request, err := http.NewRequestWithContext(ctx, http.MethodPatch, u.String(), bytes.NewReader(requestBodyBytes))
```

Comment:
nit You can encode directly to a buffer with

```suggestion
    buffer := bytes.NewBuffer(nil)
    encoder := json.NewEncoder(buffer)
    err = encoder.Encode(requestBody)
    if err != nil {
        return netip.Addr{}, fmt.Errorf("json encoding request body: %w", err)
    }

    request, err := http.NewRequestWithContext(ctx, http.MethodPatch, u.String(), buffer)
```

Reply:
Again, I just trust you 100% on this.

Changed